autobahn/asyncio/component.py

```diff
@@ -350,18 +350,24 @@ def run(components, log_level='info'):
     #   signal.signal(signal.SIGINT, signal.SIG_DFL)
 
     @asyncio.coroutine
-    def exit():
-        return loop.stop()
-
     def nicely_exit(signal):
         log.info("Shutting down due to {signal}", signal=signal)
-        for task in asyncio.Task.all_tasks():
-            task.cancel()
-        asyncio.ensure_future(exit())
+        tasks = asyncio.Task.all_tasks()
+        tasks.remove(asyncio.Task.current_task())
+        [task.cancel() for task in tasks]
+
+        try:
+            yield from asyncio.gather(*tasks)
+        except asyncio.CancelledError:
```

Comment:
(Further to @oberstet's comment) this will have to be in an inner function that returns a `Future`, and the `nicely_exit` function can return this `Future` using `add_done_callback` to add the cleanup/success handlers...and/or you could use the `txaio` helpers here (i.e. `txaio.gather` and `txaio.add_callbacks`)

"In general" it's best to write code using the `txaio` helpers in Autobahn, but in this particular case it's okay to use asyncio APIs directly (if that's easier/more-clear) because we're already in an asyncio-specific module.